modules/dists/BlockDist.chpl

```diff
@@ -869,24 +919,112 @@ iter BlockImpl.activeTargetLocales(const space : domain = boundingBox) {
   }
 }
 
-proc type Block.createDomain(dom: domain) {
-  return dom dmapped Block(dom);
+// create a domain over a Block Distribution
+proc type Block.createDomain(dom: domain, targetLocales: [] locale = Locales) {
+  return dom dmapped Block(dom, targetLocales);
+}
+
+// create a domain over a Block Distribution constructed from a series of ranges
+proc type Block.createDomain(rng: range..., targetLocales: [] locale = Locales) {
+  return createDomain({(...rng)}, targetLocales);
 }
 
 proc type Block.createDomain(rng: range...) {
   return createDomain({(...rng)});
 }
 
-proc type Block.createArray(dom: domain, type eltType) {
-  var D = createDomain(dom);
+// create an array over a Block Distribution, default initialized
+proc type Block.createArray(
+  dom: domain,
+  type eltType,
+  targetLocales: [] locale = Locales
+) {
+  var D = createDomain(dom, targetLocales);
+  var A: [D] eltType;
+  return A;
+}
+
+// create an array over a Block Distribution, initialized with the given value or iterator
+proc type Block.createArray(
+  dom: domain,
+  type eltType,
+  initExpr: ?t,
+  targetLocales: [] locale = Locales
+) where isSubtype(t, _iteratorRecord) || isCoercible(t, eltType)
+{
+  var D = createDomain(dom, targetLocales);
+  var A: [D] eltType;
+  A = initExpr;
+  return A;
+}
+
+// create an array over a Block Distribution, initialized from the given array
+proc type Block.createArray(
+  dom: domain,
+  type eltType,
+  initExpr: [?arrayDom] ?arrayEltType,
+  targetLocales: [] locale = Locales
+) where dom.rank == arrayDom.rank && isCoercible(arrayEltType, eltType)
+{
+  for (d, ad, i) in zip(dom.dims(), arrayDom.dims(), 0..) do
+    if d.size != ad.size then halt("Domain size mismatch in 'Block.createArray' dimension " + i:string);
+  var D = createDomain(dom, targetLocales);
   var A: [D] eltType;
+  A = initExpr;
   return A;
 }
 
+// create an array over a Block Distribution constructed from a series of ranges, default initialized
+proc type Block.createArray(
+  rng: range...,
+  type eltType,
+  targetLocales: [] locale = Locales
+) {
+  return createArray({(...rng)}, eltType, targetLocales);
+}
+
 proc type Block.createArray(rng: range..., type eltType) {
   return createArray({(...rng)}, eltType);
 }
 
+// create an array over a Block Distribution constructed from a series of ranges, initialized with the given value or iterator
+proc type Block.createArray(
+  rng: range...?k,
```

Comment:
I believe it should be `range(?)` above. Please check all your methods for the presence of `(?)`.

Let us also use `(?)` after `domain` uniformly in all the signatures you add or touch. Assuming the compiler accepts that.

Also if it works without the `?k` after the ellipsis, I suggest you remove it as unnecessary. In the where clauses, I prefer `rng.size == arrayDom.rank` over `k == arrayDom.rank`. I think we have made this transition elsewhere in the past.